src/batched/KokkosBatched_Gemm_Decl.hpp

```diff
@@ -95,118 +146,125 @@ struct BatchedSerialGemm;
+int BatchedGemm(const BatchedGemmHandleType *handle, const ScalarType alpha,
+                const AViewType &A, const BViewType &B, const ScalarType beta,
+                const CViewType &C) {
+  int ret = 0;
+  // Check for valid input views
+  static_assert(Kokkos::Impl::is_view<AViewType>::value,
+                "AViewType must be a Kokkos::View.");
+  static_assert(Kokkos::Impl::is_view<BViewType>::value,
+                "BViewType must be a Kokkos::View.");
+  static_assert(Kokkos::Impl::is_view<CViewType>::value,
+                "CViewType must be a Kokkos::View.");
+  static_assert(static_cast<int>(AViewType::rank) == 3,
+                "AViewType must have rank 3.");
+  static_assert(static_cast<int>(BViewType::rank) == 3,
+                "BViewType must have rank 3.");
+  static_assert(static_cast<int>(CViewType::rank) == 3,
+                "CViewType must have rank 3.");
+
+  // Check for valid data access patterns
+  // Skip checking a_layout == b_layout == c_layout
+  // Skip checking for LayoutStride
+  using c_layout = typename CViewType::array_layout;
+  if (std::is_same<c_layout, Kokkos::LayoutLeft>::value &&
+      !std::is_same<ArgBatchSzDim, BatchLayout::Right>::value) {
+    throw std::runtime_error(
+        "Error: LayoutLeft views require BatchLayout::Right");
+  }
+  if (std::is_same<c_layout, Kokkos::LayoutRight>::value &&
+      !std::is_same<ArgBatchSzDim, BatchLayout::Left>::value) {
+    throw std::runtime_error(
+        "Error: LayoutRight views require BatchLayout::Left");
+  }
 
-    // Begin checking conditions for optimal BatchedGemm invocation.
-    using view_scalar_type = typename CViewType::value_type;
-    constexpr bool is_vector =
-        KokkosBatched::is_vector<view_scalar_type>::value;
+  // Begin checking conditions for optimal BatchedGemm invocation.
+  using view_scalar_type   = typename CViewType::value_type;
+  constexpr bool is_vector = KokkosBatched::is_vector<view_scalar_type>::value;
 #if defined(KOKKOS_ENABLE_CUDA)
-    constexpr bool on_gpu =
-        std::is_same<typename CViewType::execution_space, Kokkos::Cuda>::value;
+  constexpr bool on_gpu =
+      std::is_same<typename CViewType::execution_space, Kokkos::Cuda>::value;
 #endif  // KOKKOS_ENABLE_CUDA
 #if defined(KOKKOS_ENABLE_HIP)
-    constexpr bool on_gpu = std::is_same<typename CViewType::execution_space,
-                                         Kokkos::Experimental::HIP>::value;
+  constexpr bool on_gpu = std::is_same<typename CViewType::execution_space,
+                                       Kokkos::Experimental::HIP>::value;
 #endif  // KOKKOS_ENABLE_HIP
 
 #if !defined(KOKKOS_ENABLE_CUDA) && !defined(KOKKOS_ENABLE_HIP)
-    constexpr bool on_gpu = false;
+  constexpr bool on_gpu = false;
 #endif
 
 #if __x86_64__
-    constexpr bool on_intel =
-        std::is_same<typename CViewType::execution_space::memory_space,
-                     Kokkos::HostSpace>::value;
+  constexpr bool on_intel =
```

Comment:
I am being picky here but technically, AMD is also using x86_64, so you could rename `on_intel` with `on_x86_64` but I am not sure if this was tested on AMD CPUs so maybe not worth thinking about it?